### PR TITLE
fix(react-tooltip): hide tooltips only when their trigger is fully clipped

### LINKED ISSUE #32882: [Bug]: tooltip appears outside of overflow
### Library

React Components / v9 (@fluentui/react-components)

### System Info

```shell
This repros in codepen. unable to run `npx evninfo` there.

[codesandbox demo](https://codesandbox.io/p/sandbox/old-resonance-ksvsxj?file=%2Fexample.tsx%3A10%2C32)
```

### Are you reporting an Accessibility issue?

no

### Reproduction

https://codesandbox.io/p/sandbox/old-resonance-ksvsxj?file=%2Fexample.tsx%3A10%2C32

## Actual Behavior
1. hover over a tooltip trigger element
2. scroll down the page until the tooltip should overflow into the fold of the scroll
3. the tooltip appears outside the overflow

## Expected Behavior
1. hover over a tooltip trigger element
2. scroll down the page until the tooltip should overflow into the fold of the scroll
3. the tooltip stays with the trigger anchor and appears to be hidden when scrolled out of view

### Logs

_No response_

### Requested priority

Normal

### Products/sites affected

https://copilotstudio.microsoft.com

### Are you willing to submit a PR to fix?

yes

### Validations

- [X] Check that there isn't already an issue that reports the same bug to avoid creating a duplicate.
- [X] The provided reproduction is a minimal reproducible example of the bug.

### LINKED ISSUE #36604: [Bug]: Tooltip permanently hidden inside tightly-fitted overflow:hidden containers (regression in 9.10.4+)
### Library

React Components / v9 (@fluentui/react-components)

## Actual Behavior

On `@fluentui/react-tooltip` 9.10.4+ (following #36386, which fixed #32882), a Tooltip whose trigger sits inside a tightly-fitted `overflow: hidden` container never appears at all — even at rest, with no scrolling involved.

Repro:

```jsx
<div style={{ overflow: 'hidden', display: 'flex' }}>
  <Tooltip content="I never appear on 9.10.4+" relationship="label">
    <Button>Hover me</Button>
  </Tooltip>
</div>
```

On 9.10.3 the tooltip appears as expected. On 9.10.4+ it never does.

## Expected Behavior

The tooltip should still appear when the trigger and tooltip are both within/adjacent to a tightly-fitted overflow container, as long as they aren't actually clipped/scrolled out of view.

## Root Cause

In `packages/react-components/react-positioning/library/src/usePositioningOptions.ts`, the `hide` middleware is added without a `boundary` option:

```ts
hideMiddleware({ strategy: 'referenceHidden' }),
hideMiddleware({ strategy: 'escaped' }),
```

Unlike `shift`/`flip`, which are given the same `container`/`overflowBoundary` config, `hide` falls back to floating-ui's default boundary of `'clippingAncestors'`. This means it walks up from the trigger to *any* ancestor with `overflow: hidden/scroll/auto`, not just genuine scroll containers that are actually clipping content.

In the repro above, the wrapping `<div style={{ overflow: 'hidden', display: 'flex' }}>` hugs the button tightly with no extra space above it. Since Tooltip defaults to `position="above"`, the computed tooltip position extends past that div's clip box on the very first render — no scrolling required. `escaped` becomes `true` immediately and stays that way, so the tooltip never shows (see `setHidden(escaped || referenceHidden)` in `useTooltipBase.tsx`).

## Impact

Any trigger inside a tightly-fitted `overflow: hidden` (or `scroll`/`auto`) container — flex toolbars, truncated table/list cells, card layouts, etc. — silently loses its tooltip permanently, regardless of scroll position. This is a very common CSS pattern, so this likely affects many consumers upgrading past 9.10.4.

## Proposed Fix

Pass the same `container`/`overflowBoundary` used by `shift`/`flip` into both `hideMiddleware` calls, so escape/hidden detection is scoped consistently with the rest of the positioning logic instead of picking up incidental non-scrolling clip ancestors.

### Related Issue(s)

- Regression introduced by #36386 (fix for #32882)

### Are you willing to submit a PR to fix?

yes

cc @charles-izzi

### PR DESCRIPTION
**Pull request by PaulGMardling**
## Previous Behavior

Following #36386 (which fixed #32882), a Tooltip trigger inside a tightly fitted, non-scrolling `overflow: hidden` container could remain permanently hidden. The portaled tooltip content could escape the trigger's clipping boundary even while the trigger was fully visible, causing Tooltip to hide based on Floating UI's `escaped` flag.

```jsx
<div style={{ overflow: 'hidden', display: 'flex' }}>
  <Tooltip content="I should appear" relationship="label">
    <Button>Hover me</Button>
  </Tooltip>
</div>
```

## New Behavior

Tooltip now uses only Floating UI's `referenceHidden` flag, rather than `escaped || referenceHidden`, to determine clipping-related visibility.

- Portaled tooltip content can escape the trigger's clipping ancestors without being hidden while its trigger remains visible.
- Fully clipped triggers still hide their tooltips, with visibility restored when the trigger returns into view. Partially clipped triggers do not trigger hiding.
- Escaped content alone no longer triggers hiding, including when an explicit pinned offset places the tooltip outside the viewport. This does not make offscreen content visible within the viewport.
- Positioning calculations, clipping-boundary resolution, and other positioning consumers are unchanged. No public API is added.
- The caller's `onPositioningEnd` callback continues to receive positioning events unchanged.

## Related Issues

- Fixes #36604
- Preserves hide/restore behavior for fully clipped triggers introduced for #32882

## Regression Coverage

- Unit tests cover all four combinations of `escaped` and `referenceHidden`, restoration without replacing the tooltip DOM node, and forwarding positioning events to the caller.
- Cypress coverage includes static `overflow: hidden` wrappers; partially and fully clipped triggers; scroll hide/restore; fixed-position containing blocks; dynamic overflow changes; nested clipping ancestors; all four viewport edges; pinned large offsets; and oversized tooltip content.

This section describes test coverage, not a fresh test-run result; the earlier verification counts and commands referred to a superseded implementation and have been removed.

## Change File

- `@fluentui/react-tooltip`: patch, for the Tooltip visibility regression fix.

### CODE REVIEW

**Comment by PaulGMardling** [comment]
cc @charles-izzi — this should fix the regression you found in #36604. Thanks again for the clean repro!

File: packages/react-components/react-tooltip/library/src/components/Tooltip/Tooltip.cy.tsx [hunk 1]
```diff
@@ -44,4 +44,20 @@ describe('Tooltip', () => {
         });
     });
   });
+
+  describe('static non-scrolling overflow:hidden container (regression: #36604)', () => {
```
Comment (mainframev) [inline packages/react-components/react-tooltip/library/src/components/Tooltip/Tooltip.cy.tsx]: Let's not mention issue number in the describe, remove it or add as a comment before

```suggestion
// Verifies the fix for regression reported in: https://github.com/microsoft/fluentui/issues/36604
  describe('static non-scrolling overflow:hidden container', () => {
```

File: packages/react-components/react-positioning/library/src/usePositioningOptions.ts [hunk 2]
```diff
@@ -170,8 +171,8 @@ export function usePositioningOptions(options: PositioningOptions): (
           maxSizeMiddleware(normalizedAutoSize, { container, overflowBoundary, overflowBoundaryPadding, isRtl }),
         intersectingMiddleware(),
         arrow && arrowMiddleware({ element: arrow, padding: arrowPadding }),
-        hideMiddleware({ strategy: 'referenceHidden' }),
-        hideMiddleware({ strategy: 'escaped' }),
+        hideMiddleware({ strategy: 'referenceHidden', hasScrollableElement }),
+        hideMiddleware({ strategy: 'escaped', hasScrollableElement }),
```
Comment (mainframev) [inline packages/react-components/react-positioning/library/src/usePositioningOptions.ts]: @PaulGMardling  

q: do we really need a custom wrapper around floating-ui hide middleware? won't be better to inline that instead of introducing additional wrapper? 

```ts
  hideMiddleware({ strategy: 'escaped', boundary: hasScrollableElement ? 'clippingAncestors' : [] }),
```

File: packages/react-components/react-positioning/library/src/middleware/hide.ts [hunk 3]
```diff
@@ -0,0 +1,43 @@
+import type { Middleware } from '@floating-ui/dom';
+import { hide as baseHide } from '@floating-ui/dom';
+
+export interface HideMiddlewareOptions {
```
Comment (mainframev) [inline packages/react-components/react-positioning/library/src/middleware/hide.ts]: I think you can import and extend the type, instead of introducing a a completely new one 

```ts
import type { HideOptions } from '@floating-ui/dom';
```

**Review by mainframev — COMMENTED** [review]
silently reverts the previous fix (I assume because of the `boundary: []` the trigger scrolled out of a scroll container is no longer detected), visible in currently deployed SB: 

https://github.com/user-attachments/assets/de4149cf-bf7f-4aae-891f-f8204607dfda

File: packages/react-components/react-positioning/library/src/middleware/hide.test.ts [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+ * (both the pre-existing scroll-based test for #32882, and the new static `overflow: hidden` test
+ * for #36604) — a real browser is required to reliably exercise floating-ui's
+ * offset-parent/scale/layout math; jsdom's emulation of that math doesn't match real browser
+ * behavior closely enough to pin it in a unit test here.
+ *
+ * These tests instead pin the narrower, fully deterministic thing this module is responsible for:
+ * that it maps `hasScrollableElement` to the correct `boundary` option passed to the underlying
+ * `@floating-ui/dom` `hide` middleware.
+ */
+describe('hide', () => {
```
Comment (mainframev) [inline packages/react-components/react-positioning/library/src/middleware/hide.test.ts]: I doubt this test is useful and can be omitted, related to this comment: https://github.com/microsoft/fluentui/pull/36605/changes#r3836947174

File: packages/react-components/react-positioning/library/src/usePositioningOptions.ts [hunk 5]
```diff
@@ -170,8 +171,8 @@ export function usePositioningOptions(options: PositioningOptions): (
           maxSizeMiddleware(normalizedAutoSize, { container, overflowBoundary, overflowBoundaryPadding, isRtl }),
         intersectingMiddleware(),
         arrow && arrowMiddleware({ element: arrow, padding: arrowPadding }),
-        hideMiddleware({ strategy: 'referenceHidden' }),
-        hideMiddleware({ strategy: 'escaped' }),
```
Comment (mainframev) [inline packages/react-components/react-positioning/library/src/usePositioningOptions.ts:174]: I feel it can be risky: that applies to every consumer of `usePositioning` - `react-combobox`,
`react-datepicker-compat`, `react-menu`, `react-popover`, `react-tag-picker`, `react-tooltip`, so quite large blast radius. `escaped`/`referenceHidden` go straight out as `data-popper-escaped` / `data-popper-reference-hidden`

Wdyt about adding it via a new react-positioning option `hideBoundary` and enabling it for Tooltip
only? 

```ts
const positioningOptions = {
  enabled: state.visible,
  arrowPadding: 2 * tooltipBorderRadius,
  position: 'above' as const,
  align: 'center' as const,
  offset: 4,
  hideBoundary: 'scrollParent' as const,
  ...resolvedPositioning,
  onPositioningEnd,
};
```

File: packages/react-components/react-positioning/library/src/middleware/hide.test.ts (hunk 4 above)
Comment (PaulGMardling) [reply packages/react-components/react-positioning/library/src/middleware/hide.test.ts]: I'll remove it, good catch

File: packages/react-components/react-tooltip/library/src/components/Tooltip/Tooltip.cy.tsx (hunk 1 above)
Comment (PaulGMardling) [reply packages/react-components/react-tooltip/library/src/components/Tooltip/Tooltip.cy.tsx]: Thank you helpful information certainly

File: packages/react-components/react-positioning/library/src/usePositioningOptions.ts (hunk 5 above)
Comment (PaulGMardling) [reply packages/react-components/react-positioning/library/src/usePositioningOptions.ts:174]: That definitely removes the blast radius! Thanks for the suggestion

**Comment by PaulGMardling** [comment]
> silently reverts the previous fix (I assume because of the `boundary: []` the trigger scrolled out of a scroll container is no longer detected), visible in currently deployed SB:
> 
>  Screen.Recording.2026-08-22.at.22.16.07.mov

Fixed

https://github.com/user-attachments/assets/7110d073-aa63-436b-8341-7f672d0e4f0a


https://github.com/user-attachments/assets/c022a2d6-5837-4cd9-acd3-3339313f5ccd

File: packages/react-components/react-positioning/library/src/middleware/hide.ts (hunk 3 above)
Comment (PaulGMardling) [reply packages/react-components/react-positioning/library/src/middleware/hide.ts]: I removed the middleware extension altogether and inlined it, much cleaner solution, thanks

File: packages/react-components/react-positioning/library/src/usePositioningOptions.ts (hunk 2 above)
Comment (PaulGMardling) [reply packages/react-components/react-positioning/library/src/usePositioningOptions.ts]: No, we don't, too heavy of a solution for the single component use case, thanks again

File: packages/react-components/react-tooltip/library/src/components/Tooltip/TooltipOverflowHidden.test.tsx [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+      // The tooltip bubble itself (and anything else, e.g. the arrow): give it a modest,
+      // non-zero size so Floating UI's placement math has something realistic to work with.
+      return tooltipContentRect;
+    }
+
+    // JSDOM has no real layout engine: `getBoundingClientRect` always returns a zero rect, and
+    // `offsetWidth`/`offsetHeight` (which Floating UI's dimension measurement actually reads via
+    // `getCssDimensions`) are always 0 too. Both need mocking in lockstep, or Floating UI falls
+    // back to treating every element as zero-size regardless of the rects above.
+    spies = [
```
Comment (dmytrokirpa) [inline packages/react-components/react-tooltip/library/src/components/Tooltip/TooltipOverflowHidden.test.tsx]: that's a problem, with these changes all cosnumers' tests need to use spies/mocks for `getBoundingClientRect`

Comment (PaulGMardling) [reply packages/react-components/react-tooltip/library/src/components/Tooltip/TooltipOverflowHidden.test.tsx]: I removed it

**Comment by PaulGMardling** [comment]
test from https://github.com/microsoft/fluentui/pull/36689 were integrated, code updated and storybook checked to verify edge cases still pass

**Comment by PaulGMardling** [comment]
@bsunderhus 
Any chance you can review this?

File: packages/react-components/react-tooltip/stories/src/Tooltip/TooltipStaticOverflowHidden.stories.tsx [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+    borderRadius: tokens.borderRadiusMedium,
+    padding: tokens.spacingVerticalXS,
+  },
+});
+
+export const StaticOverflowHidden = (): JSXElement => {
+  const styles = useStyles();
+
+  return (
+    <div className={styles.root}>
```
Comment (dmytrokirpa) [inline packages/react-components/react-tooltip/stories/src/Tooltip/TooltipStaticOverflowHidden.stories.tsx]: q: should this be a e2e tests rather then a documentation example?

Comment (PaulGMardling) [reply packages/react-components/react-tooltip/stories/src/Tooltip/TooltipStaticOverflowHidden.stories.tsx]: Yes — I moved the regression coverage to Cypress e2e and removed the documentation-only Storybook example. The e2e tests now verify both the static  overflow: hidden  case and hiding when an outer nested scroll container clips the trigger.

**Comment by bsunderhus** [comment]
Related https://github.com/microsoft/fluentui/pull/36758

**Review by bsunderhus — APPROVED** [review]

**Review by mainframev — APPROVED** [review]
